webapp/packages/core-root/src/AsyncTask/AsyncTaskInfoService.ts

```diff
@@ -0,0 +1,155 @@
+/*
+ * CloudBeaver - Cloud Database Manager
+ * Copyright (C) 2020-2024 DBeaver Corp and others
+ *
+ * Licensed under the Apache License, Version 2.0.
+ * you may not use this file except in compliance with the License.
+ */
+import type { Subscription } from 'rxjs';
+
+import { Disposable, injectable } from '@cloudbeaver/core-di';
+import { type AsyncTaskInfo, GraphQLService, type WsAsyncTaskInfo } from '@cloudbeaver/core-sdk';
+
+import type { Unsubscribe } from '../ServerEventEmitter/IServerEventEmitter.js';
+import { ServerEventId } from '../SessionEventSource.js';
+import { AsyncTask } from './AsyncTask.js';
+import { AsyncTaskInfoEventHandler } from './AsyncTaskInfoEventHandler.js';
+
+@injectable()
+export class AsyncTaskInfoService extends Disposable {
+  private readonly tasks: Map<string, AsyncTask>;
+  private readonly taskIdAliases: Map<string, string>;
+  private readonly pendingEvents: Map<string, WsAsyncTaskInfo>;
+  private connection: Subscription | null;
+  private onEventUnsubscribe: Unsubscribe | null;
+
+  constructor(
+    private readonly graphQLService: GraphQLService,
+    private readonly asyncTaskInfoEventHandler: AsyncTaskInfoEventHandler,
+  ) {
+    super();
+    this.tasks = new Map();
+    this.taskIdAliases = new Map();
+    this.pendingEvents = new Map();
+    this.connection = null;
+    this.handleEvent = this.handleEvent.bind(this);
+
+    this.onEventUnsubscribe = asyncTaskInfoEventHandler.onEvent<WsAsyncTaskInfo>(ServerEventId.CbSessionTaskInfoUpdated, this.handleEvent);
+  }
+
+  private async updateTask(task: AsyncTask, data: WsAsyncTaskInfo) {
+    if (data.running === false) {
+      await task.updateInfoAsync(async () => {
+        const { taskInfo } = await this.graphQLService.sdk.getAsyncTaskInfo({
+          taskId: data.taskId,
+          removeOnFinish: false,
+        });
+
+        return taskInfo;
+      });
+    } else {
+      task.updateStatus(data);
+    }
+  }
+
+  private async handleEvent(data: WsAsyncTaskInfo) {
+    const task = this.getTask(data.taskId);
+
+    if (!task) {
+      this.pendingEvents.set(data.taskId, data);
+      return;
+    }
+
+    await this.updateTask(task, data);
+  }
+
+  override dispose(): void {
+    this.connection?.unsubscribe();
+    this.onEventUnsubscribe?.();
+  }
+
+  create(getter: () => Promise<AsyncTaskInfo>): AsyncTask {
+    const task = new AsyncTask(getter, this.cancelTask.bind(this));
+
+    this.tasks.set(task.id, task);
+    task.onStatusChange.addHandler(info => {
+      const pendingEvent = this.pendingEvents.get(info.id);
+      if (pendingEvent) {
+        const changes = { ...pendingEvent };
+        this.pendingEvents.delete(info.id);
+        this.updateTask(task, changes);
+      }
+      if (!this.taskIdAliases.get(info.id)) {
+        this.taskIdAliases.set(info.id, task.id);
+      }
```

Comment:
it's better to isolate this logic to "only first time"
1. check if we have task alias already, if we do: do nothing
2. else set task alias, then check for pending events and process if we have some